pkg/controller/component/service_descriptor_utils.go

```diff
@@ -231,34 +247,37 @@ func referencedVars4Legacy(ctx context.Context, cli client.Reader, namespace str
 		Name: constant.GenerateDefaultConnCredential(serviceRef.Cluster),
 	}
 	if err := cli.Get(ctx, secretKey, secret); err != nil {
-		return nil, err
+		return err
 	}
 
-	vars := []*appsv1alpha1.CredentialVar{nil, nil, nil, nil, nil}
-	keys := []string{
-		constant.ServiceDescriptorEndpointKey,
-		constant.ServiceDescriptorHostKey,
-		constant.ServiceDescriptorPortKey,
-		constant.ServiceDescriptorUsernameKey,
-		constant.ServiceDescriptorPasswordKey,
+	copySecretDataToCredentialVar("", secret, constant.ServiceDescriptorEndpointKey, &vars.endpoint)
+	copySecretDataToCredentialVar("", secret, constant.ServiceDescriptorPortKey, &vars.port)
+	copySecretDataToCredentialVar("", secret, constant.ServiceDescriptorUsernameKey, &vars.username)
+	copySecretDataToCredentialVar("", secret, constant.ServiceDescriptorPasswordKey, &vars.password)
+
+	// don't set the host and podFQDNs for legacy clusters
+	vars.host = nil
+	vars.podFQDNs = nil
+
+	return nil
+}
+
+func copySecretDataToCredentialVar(namespace string, secret *corev1.Secret, key string, v **appsv1alpha1.CredentialVar) {
+	if _, ok := secret.Data[key]; !ok {
+		return
 	}
-	for idx, key := range keys {
-		if key == constant.ServiceDescriptorHostKey {
-			// don't set the host for legacy clusters
-			continue
-		}
-		if _, ok := secret.Data[key]; ok {
-			vars[idx] = &appsv1alpha1.CredentialVar{
-				ValueFrom: &corev1.EnvVarSource{
-					SecretKeyRef: &corev1.SecretKeySelector{
-						LocalObjectReference: corev1.LocalObjectReference{Name: secret.Name},
-						Key:                  key,
-					},
+	if secret.Namespace == namespace || namespace == "" {
+		*v = &appsv1alpha1.CredentialVar{
+			ValueFrom: &corev1.EnvVarSource{
+				SecretKeyRef: &corev1.SecretKeySelector{
+					LocalObjectReference: corev1.LocalObjectReference{Name: secret.Name},
+					Key:                  key,
 				},
-			}
+			},
 		}
+	} else {
+		*v = &appsv1alpha1.CredentialVar{Value: string(secret.Data[key])}
 	}
```

Comment:
t's best to discard this logic and not support referenced secrets that are not in the same namespace for security reasons.

Reply:
Disabled on the new API.